test/Table_Tests/src/Expression_Spec.enso

```diff
@@ -0,0 +1,220 @@
+        specify_test "should be able to add a date or time column" expression_test->
+            expression_test "#2020-12-23#" (Date.new 2020 12 23)
+            expression_test "#12:34#" (Time_Of_Day.new 12 34)
+            expression_test "#12:34:56#" (Time_Of_Day.new 12 34 56)
+            expression_test "#12:34:56.789#" (Time_Of_Day.new 12 34 56 789000000)
+            expression_test "#12:34:56.789000123#" (Time_Of_Day.new 12 34 56 789000123)
+            expression_test "#2020-12-23 12:34#" (Date_Time.new 2020 12 23 12 34)
+            expression_test "#2020-12-23 12:34:56#" (Date_Time.new 2020 12 23 12 34 56)
+            expression_test "#2020-12-23 12:34:56+02:30#" (Date_Time.new 2020 12 23 10 04 56)
+            expression_test "#2020-12-23 12:34:56Z[UTC]#" (Date_Time.new 2020 12 23 12 34 56 zone=Time_Zone.utc)
+            expression_test "#2020-12-23 12:34:56.157+01[UTC]#" (Date_Time.new 2020 12 23 11 34 56 157000000 zone=Time_Zone.utc)
+            expression_test "#2020-12-23T12:34[Europe/Warsaw]#" (Date_Time.new 2020 12 23 12 34 zone=Time_Zone.parse("Europe/Warsaw"))
+
+    Test.group "Expression Arithmetic" <|
+        specify_test "should be able to do basic arithmetic" expression_test->
+            expression_test "1+1" 2
+            expression_test "23-15" 8
+            expression_test "2.5*4.2" 10.5
+            expression_test "1_000.456/2" 500.228
+            expression_test "2^4" 16
+            expression_test "11%3" 2
+
+        specify_test "should be able to do basic arithmetic with order" expression_test->
+            expression_test "1+1*2+2" 5
+            expression_test "23-15/3+6" 24
+            expression_test "52.92/4.2^2" 3
+            expression_test "(1+1)*2+2" 6
+
+        specify_test "should be able to do basic arithmetic with whitespace" expression_test->
+            expression_test "1 + 1" 2
+            expression_test "  23 -15  " 8
+            expression_test "2.5*  4.2" 10.5
+            expression_test "1_000.456/   2" 500.228
+            expression_test " 2 ^ 4 " 16
+            expression_test "  11   % 3    " 2
+
+    Test.group "Column Arithmetic" <|
+        specify_test "should be able to perform arithmetic on columns" expression_test->
+            expression_test "[A] + 2" [3, 4, 5, 6, 7]
+            expression_test "[B] - 2" [-1, -0.5, 0.5, 2, 4]
+            expression_test "[A] * 4" [4, 8, 12, 16, 20]
+            expression_test "[B] / 2" [0.5, 0.75, 1.25, 2, 3]
+            expression_test "[A] + [B]" [2, 3.5, 5.5, 8, 11]
+            expression_test "[A] - [B]" [0, 0.5, 0.5, 0, -1]
+            expression_test "[A] * [B]" [1, 3, 7.5, 16, 30]
+            expression_test "[B] / [A]" [1, 0.75, 0.8333333333333334, 1, 1.2]
+            expression_test "[A] ^ [B]" [1, 2.8284271247461903, 15.588457268119896, 256, 15625]
+            expression_test "[A] % [B]" [0, 0.5, 0.5, 0, 5]
+            expression_test "[A] + [B] + [A]" [3, 5.5, 8.5, 12, 16]
```

Comment:
Ten years ago I had a [DSL vs. API design shootout](http://wiki.apidesign.org/wiki/DSL) with a friend at a conference. We had to [write our notes down](http://wiki.apidesign.org/wiki/DSL) to even understand each other. Given our classification, what we have here is an **external DSL** with its **own parser** _embedded in an Enso string_.

A string based DSL approach will increase the complexity of our tooling. We will not be able to benefit from Enso type system, use Enso suggestions DB, etc. We will have to find ways for the IDE to understand this DSL syntax.

Given the sole purpose of Enso was to work effectively with tables, this feels as a _side step_. A recognition that Enso isn't good enough at manipulating tables yet. I sincerely hope that over time Enso improves and provides good enough framework for dealing with tables as well.

Ideally we should create in **internal DSL**, reuse the **Enso parser**,  _intermix the DSL_ in between other Enso constructs. Make sure the Enso _type system straightens_ the DSL and the DSL is 1st class citizen in the Enso code and can be naturally manipulated by the IDE.